### PR TITLE
🌱 Deduplicate AdoptMachinePorts and AdoptBastionPorts

### PR DESCRIPTION
**Pull request by mdbooth**
Both of these methods rely on ReferencedMachineResources and DependentMachineResources, so they can be easily refactored to have a common implementation.

Fixes: #1942 

This PR is based on #1941, so we should merge that first.

TODO:
- [x] Merge #1941 

/hold

### CODE REVIEW

File: pkg/cloud/services/compute/dependent_resources_test.go
Comment (EmilienM) [inline pkg/cloud/services/compute/dependent_resources_test.go:1]: can't we keep the tests or did they become useless?

Comment (mdbooth) [reply pkg/cloud/services/compute/dependent_resources_test.go:1]: The function is such a thin wrapper round Adopt() I didn't think it was worth of its own unit tests. They're all basically adoption tests. You don't agree?

Comment (mdbooth) [reply pkg/cloud/services/compute/dependent_resources_test.go:1]: Ok, it looks like these were the Adopt tests. I'll resurrect them in port_test.go.

Comment (EmilienM) [reply pkg/cloud/services/compute/dependent_resources_test.go:1]: Thanks.

Comment (mdbooth) [reply pkg/cloud/services/compute/dependent_resources_test.go:1]: Bah, forgot. Don't let me forget.

Comment (mdbooth) [reply pkg/cloud/services/compute/dependent_resources_test.go:1]: I moved them into port_test.go called Test_AdoptPorts, and added a few more.

**Comment by mdbooth** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1954 was the root cause of most of the full-test CI failures in this PR. This PR includes a change, which we needed to make anyway, which avoids the issue in most cases: we don't need to re-reconcile if we only adopted. However, the underlying issue remains. I'll push a separate PR to fix that.

File: pkg/cloud/services/compute/dependent_resources.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ import (
 func AdoptDependentMachineResources(scope *scope.WithLogger, baseName string, referencedResources *infrav1.ReferencedMachineResources, dependentResources *infrav1.DependentMachineResources) error {
 	networkingService, err := networking.NewService(scope)
 	if err != nil {
-		return false, err
```
Comment (dulek) [inline pkg/cloud/services/compute/dependent_resources.go]: Seems like this change should be in a different commit? It's syntax error otherwise.

**Review by dulek — CHANGES_REQUESTED** [review]

**Comment by mdbooth** [comment]
> Alright, I only found one issue with a change done in the last commit, when it should be done in the middle one. Let's fix that so that reverts would be possible? Or just squash the commits.

Thanks. I'll tidy them up but keep them separate, as I've been *trying* to make smaller, functional commits. When you look at the next one in this series, though, you'll see how that's going 😟

**Comment by mdbooth** [comment]
I've cleaned that up and re-run the non-e2e tests against each commit individually. I also rebased on to main to pick up #1955, which is important for this series.

**Pull request merged**